packages/core/contracts/chainbridge/BeaconOracle.sol

```diff
@@ -0,0 +1,164 @@
+ * from the Source Oracle and makes it available on non-Mainnet chains. The "Sink" Oracle can also be used to trigger
+ * price requests from the DVM on Mainnet.
+ */
+abstract contract BeaconOracle is OracleAncillaryInterface {
+    enum RequestState { NeverRequested, Requested, Resolved }
+
+    struct Price {
+        RequestState state;
+        int256 price;
+    }
+
+    // Chain ID for this Oracle.
+    uint8 public currentChainID;
+
+    // Mapping of encoded price requests {identifier, time, ancillaryData} to Price objects.
+    mapping(bytes32 => Price) internal prices;
+
+    // Finder to provide addresses for DVM system contracts.
+    FinderInterface public finder;
+
+    event PriceRequestAdded(
+        address indexed requester,
+        uint8 indexed chainID,
+        bytes32 indexed identifier,
+        uint256 time,
+        bytes ancillaryData
+    );
+    event PushedPrice(
+        address indexed pusher,
+        uint8 indexed chainID,
+        bytes32 indexed identifier,
+        uint256 time,
+        bytes ancillaryData,
+        int256 price
+    );
+
+    /**
+     * @notice Constructor.
+     * @param _finderAddress finder to use to get addresses of DVM contracts.
+     */
+    constructor(address _finderAddress, uint8 _chainID) public {
+        finder = FinderInterface(_finderAddress);
+        currentChainID = _chainID;
+    }
+
+    // We assume that there is only one GenericHandler for this network.
+    modifier onlyGenericHandlerContract() {
+        require(
+            msg.sender == finder.getImplementationAddress(OracleInterfaces.GenericHandler),
+            "Caller must be GenericHandler"
+        );
+        _;
+    }
+
+    // This assumes that the local network has a Registry that resembles the Mainnet registry.
+    modifier onlyRegisteredContract() {
+        RegistryInterface registry = RegistryInterface(finder.getImplementationAddress(OracleInterfaces.Registry));
+        require(registry.isContractRegistered(msg.sender), "Caller must be registered");
+        _;
+    }
+
+    /**
+     * @notice Returns whether a price has resolved for the request.
+     * @return True if a price is available, False otherwise. If true, then getPrice will succeed for the request.
+     */
+
+    function hasPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public view override onlyRegisteredContract() returns (bool) {
+        bytes32 priceRequestId = _encodePriceRequest(currentChainID, identifier, time, ancillaryData);
+        return prices[priceRequestId].state == RequestState.Resolved;
+    }
+
+    /**
+     * @notice Returns resolved price for the request.
+     * @return int256 Price, or reverts if no resolved price for any reason.
+     */
+
+    function getPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public view override onlyRegisteredContract() returns (int256) {
```

Comment:
Just to double check -- we would never really use these functions on the source side right? Because the source can't request prices for itself, right?

Reply:
well I suppose you could call this from the source side, but it'd be only worst than calling the DVM directly.